source/loader/ur_adapter_registry.hpp

```diff
@@ -88,9 +82,54 @@ class AdapterRegistry {
     static constexpr std::array<const char *, 1> knownPlatformNames{
         MAKE_LIBRARY_NAME("ur_adapter_level_zero", "0")};
 
+    std::optional<std::vector<fs::path>> getEnvAdapterSearchPaths() {
+        std::vector<fs::path> paths;
+        auto pathStringsOpt = getenv_to_vec("UR_ADAPTERS_SEARCH_PATH");
+        if (pathStringsOpt.has_value()) {
+            auto pathStrings = pathStringsOpt.value();
+            std::transform(pathStrings.cbegin(), pathStrings.cend(),
+                           std::back_inserter(paths), [](const std::string &s) {
+                               return fs::absolute(fs::path(s));
+                           });
+            return paths;
+        }
+
+        return std::nullopt;
+    }
+
     void discoverKnownAdapters() {
-        for (const auto &path : knownPlatformNames) {
-            discovered_adapters.emplace_back(path);
+        std::vector<fs::path> searchPaths;
+
+        auto searchPathsEnvOpt = getEnvAdapterSearchPaths();
+        if (searchPathsEnvOpt.has_value()) {
+            auto searchPathsEnv = searchPathsEnvOpt.value();
+            std::copy(searchPathsEnv.begin(), searchPathsEnv.end(),
+                      std::back_inserter(searchPaths));
+        }
+
+        auto searchPathsOSOpt = getOSSearchPaths();
```

Comment:
I'm actually not sure we should have the same search order for both linux and Windows. On Windows, `PATH` is expected to be checked last I think (after the current directory): https://learn.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order?redirectedfrom=MSDN#search_order_for_desktop_applications

Reply:
I've grouped the current directory, environmental variable paths, and system directory under the `getOSSearchPaths(...)` function and resolved the search order there.